### ISSUE TITLE
Nullness checker fails to infer nonnullness of variable

### ISSUE THREAD

**Issue by anbenson** [typetools/checker-framework#1870]
Stumbled across this at work, and was able to whittle it to a small test case.

Test case (as a txt since Github won't let me upload a java file, but it's java):
[BugDemo.txt](https://github.com/typetools/checker-framework/files/1782787/BugDemo.txt)

Output of ./checker-framework-2.4.0/checker/bin/javac -processor nullness -version -verbose -AprintErrorStack -AprintAllQualifiers BugDemo.java:
[verboseOutput.txt](https://github.com/typetools/checker-framework/files/1782791/verboseOutput.txt)

Test case:

```java
import org.checkerframework.checker.nullness.qual.Nullable;
class BugFactory {
  Bug create() {
    return new Bug();
  }
}
class Bug {
  void buzz(String s) {
    // bzzz
  }
}
class BugDemo {
  @Nullable private String myString;
  void run() {
    BugFactory bugFactory = new BugFactory();
    if (myString != null) {
      Bug bug = bugFactory.create();
      bug.buzz(myString);
    }
  }
}
```

As you can see, I explicitly check that myString is not null before passing it to bug.buzz(). However, the nullness checker believes it is still possibly null:

```02:11 PM $ ./checker-framework-2.4.0/checker/bin/javac -processor nullness BugDemo.java 
BugDemo.java:22: error: [argument.type.incompatible] incompatible types in argument.
      bug.buzz(myString);
               ^
  found   : @Initialized @Nullable String
  required: @Initialized @NonNull String
1 error
```

If I change the previous line to read

```java
Bug bug = new Bug();
```

then the nullness checker is happy (so yes, all the extra classes and indirection are necessary).

Another workaround is to move the previous line, unmodified, to before my if-check. So I can workaround this, but whenever I touch this class (or the class that inspired this, I don't work with BugFactory's at work) I occasionally trigger this bug, leading to confusion :)

**Comment by smillst** [comment]
`bugFactory.create()` could set `myString` to null.  For example,
```
class BugFactory {
    BugDemo demo;
    BugFactory(BugDemo demo) {
        this.demo = demo;
    }
    Bug create() {
        demo.myString = null;  // Or use some setter to set myString to null
        return new Bug();
    }
}
class Bug {
    public  Bug(){

    }
    void buzz(String s) {
        // bzzz
    }
}
class BugDemo {
    @Nullable private String myString;
    void run() {
        BugFactory bugFactory = new BugFactory(this);
        if (myString != null) {
            Bug bug = bugFactory.create();
            bug.buzz(myString);
        }
    }
}
```
Because of this, after the call to `create()`, `myString` is nullable.  See https://checkerframework.org/manual/#type-refinement-fields for more details. Also, 
 https://checkerframework.org/manual/#type-refinement-purity explains how to annotated methods so that fields don't loss refinement after they are called.

**Comment by anbenson** [comment]
That's helpful, thank you!
